// @nomadreservations/ngx-gallery/esm2015/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of gallery
 */
export { NgxGalleryActionComponent } from './lib/ngx-gallery-action.component';
export { NgxGalleryAnimation } from './lib/ngx-gallery-animation.model';
export { NgxGalleryArrowsComponent } from './lib/ngx-gallery-arrows.component';
export { NgxGalleryBulletsComponent } from './lib/ngx-gallery-bullets.component';
export { NgxGalleryImageSize } from './lib/ngx-gallery-image-size.model';
export { NgxGalleryImageComponent } from './lib/ngx-gallery-image.component';
export { NgxGalleryImage } from './lib/ngx-gallery-image.model';
export { NgxGalleryLayout } from './lib/ngx-gallery-layout.model';
export { NgxGalleryOptions } from './lib/ngx-gallery-options.model';
export { NgxGalleryOrder } from './lib/ngx-gallery-order.model';
export { NgxGalleryOrderedImage } from './lib/ngx-gallery-ordered-image.model';
export { NgxGalleryPreviewComponent } from './lib/ngx-gallery-preview.component';
export { NgxGalleryThumbnailsComponent } from './lib/ngx-gallery-thumbnails.component';
export { NgxGalleryComponent } from './lib/ngx-gallery.component';
export { CustomHammerConfig, NgxGalleryModule, NgxGalleryAction, NgxGalleryHelperService } from './lib/ngx-gallery.module';
