// @nomadreservations/ngx-gallery/esm2015/lib/ngx-gallery.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { HammerGestureConfig, HAMMER_GESTURE_CONFIG } from '@angular/platform-browser';
import { NgxGalleryActionComponent } from './ngx-gallery-action.component';
import { NgxGalleryArrowsComponent } from './ngx-gallery-arrows.component';
import { NgxGalleryBulletsComponent } from './ngx-gallery-bullets.component';
import { NgxGalleryImageComponent } from './ngx-gallery-image.component';
import { NgxGalleryPreviewComponent } from './ngx-gallery-preview.component';
import { NgxGalleryThumbnailsComponent } from './ngx-gallery-thumbnails.component';
import { NgxGalleryComponent } from './ngx-gallery.component';
export { NgxGalleryActionComponent } from './ngx-gallery-action.component';
export { NgxGalleryAction } from './ngx-gallery-action.model';
export { NgxGalleryAnimation } from './ngx-gallery-animation.model';
export { NgxGalleryArrowsComponent } from './ngx-gallery-arrows.component';
export { NgxGalleryBulletsComponent } from './ngx-gallery-bullets.component';
export { NgxGalleryHelperService } from './ngx-gallery-helper.service';
export { NgxGalleryImageSize } from './ngx-gallery-image-size.model';
export { NgxGalleryImageComponent } from './ngx-gallery-image.component';
export { NgxGalleryImage } from './ngx-gallery-image.model';
export { NgxGalleryLayout } from './ngx-gallery-layout.model';
export { NgxGalleryOptions } from './ngx-gallery-options.model';
export { NgxGalleryOrder } from './ngx-gallery-order.model';
export { NgxGalleryOrderedImage } from './ngx-gallery-ordered-image.model';
export { NgxGalleryPreviewComponent } from './ngx-gallery-preview.component';
export { NgxGalleryThumbnailsComponent } from './ngx-gallery-thumbnails.component';
export { NgxGalleryComponent } from './ngx-gallery.component';
export class CustomHammerConfig extends HammerGestureConfig {
    constructor() {
        super(...arguments);
        this.overrides = (/** @type {?} */ ({
            pinch: { enable: false },
            rotate: { enable: false }
        }));
    }
}
if (false) {
    /** @type {?} */
    CustomHammerConfig.prototype.overrides;
}
export class NgxGalleryModule {
}
NgxGalleryModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                declarations: [
                    NgxGalleryActionComponent,
                    NgxGalleryArrowsComponent,
                    NgxGalleryBulletsComponent,
                    NgxGalleryImageComponent,
                    NgxGalleryThumbnailsComponent,
                    NgxGalleryPreviewComponent,
                    NgxGalleryComponent
                ],
                exports: [NgxGalleryComponent],
                providers: [{ provide: HAMMER_GESTURE_CONFIG, useClass: CustomHammerConfig }]
            },] }
];
