// @nomadreservations/ngx-gallery/esm2015/lib/ngx-gallery-action.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function INgxGalleryAction() { }
if (false) {
    /** @type {?} */
    INgxGalleryAction.prototype.icon;
    /** @type {?|undefined} */
    INgxGalleryAction.prototype.disabled;
    /** @type {?|undefined} */
    INgxGalleryAction.prototype.titleText;
    /** @type {?} */
    INgxGalleryAction.prototype.onClick;
}
export class NgxGalleryAction {
    /**
     * @param {?} action
     */
    constructor(action) {
        this.icon = action.icon;
        this.disabled = action.disabled ? action.disabled : false;
        this.titleText = action.titleText ? action.titleText : '';
        this.onClick = action.onClick;
    }
}
if (false) {
    /** @type {?} */
    NgxGalleryAction.prototype.icon;
    /** @type {?} */
    NgxGalleryAction.prototype.disabled;
    /** @type {?} */
    NgxGalleryAction.prototype.titleText;
    /** @type {?} */
    NgxGalleryAction.prototype.onClick;
}
