// @nomadreservations/ngx-gallery/esm2015/lib/ngx-gallery-options.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgxGalleryAnimation } from './ngx-gallery-animation.model';
import { NgxGalleryImageSize } from './ngx-gallery-image-size.model';
import { NgxGalleryLayout } from './ngx-gallery-layout.model';
import { NgxGalleryOrder } from './ngx-gallery-order.model';
import { NgxGalleryAction } from './ngx-gallery-action.model';
/**
 * @record
 */
export function INgxGalleryOptions() { }
if (false) {
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.width;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.height;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.breakpoint;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.fullWidth;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.layout;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.startIndex;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.linkTarget;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.lazyLoading;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.image;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imagePercent;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageArrows;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageArrowsAutoHide;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageSwipe;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageAnimation;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageSize;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageAutoPlay;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageAutoPlayInterval;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageAutoPlayPauseOnHover;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageInfinityMove;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageActions;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageDescription;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.imageBullets;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnails;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsColumns;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsRows;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsPercent;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsMargin;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsArrows;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsArrowsAutoHide;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsSwipe;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsMoveSize;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsOrder;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsRemainingCount;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsAsLinks;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailsAutoHide;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailMargin;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailSize;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.thumbnailActions;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.preview;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewDescription;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewArrows;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewArrowsAutoHide;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewSwipe;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewFullscreen;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewForceFullscreen;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewCloseOnClick;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewCloseOnEsc;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewKeyboardNavigation;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewAnimation;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewAutoPlay;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewAutoPlayInterval;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewAutoPlayPauseOnHover;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewInfinityMove;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewZoom;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewZoomStep;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewZoomMax;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewZoomMin;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewRotate;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewDownload;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewCustom;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.previewBullets;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.arrowPrevIcon;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.arrowNextIcon;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.closeIcon;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.fullscreenIcon;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.spinnerIcon;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.zoomInIcon;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.zoomOutIcon;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.rotateLeftIcon;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.rotateRightIcon;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.downloadIcon;
    /** @type {?|undefined} */
    INgxGalleryOptions.prototype.actions;
}
export class NgxGalleryOptions {
    /**
     * @param {?} obj
     */
    constructor(obj) {
        /** @type {?} */
        const preventDefaults = obj.breakpoint === undefined ? false : true;
        /**
         * @template T
         * @param {?} source
         * @param {?} defaultValue
         * @return {?}
         */
        function use(source, defaultValue) {
            return obj && (source !== undefined || preventDefaults) ? source : defaultValue;
        }
        this.breakpoint = use(obj.breakpoint, undefined);
        this.width = use(obj.width, '500px');
        this.height = use(obj.height, '400px');
        this.fullWidth = use(obj.fullWidth, false);
        this.layout = use(obj.layout, NgxGalleryLayout.ThumbnailsBottom);
        this.startIndex = use(obj.startIndex, 0);
        this.linkTarget = use(obj.linkTarget, '_blank');
        this.lazyLoading = use(obj.lazyLoading, true);
        this.image = use(obj.image, true);
        this.imagePercent = use(obj.imagePercent, 75);
        this.imageArrows = use(obj.imageArrows, true);
        this.imageArrowsAutoHide = use(obj.imageArrowsAutoHide, false);
        this.imageSwipe = use(obj.imageSwipe, false);
        this.imageAnimation = use(obj.imageAnimation, NgxGalleryAnimation.Fade);
        this.imageSize = use(obj.imageSize, NgxGalleryImageSize.Cover);
        this.imageAutoPlay = use(obj.imageAutoPlay, false);
        this.imageAutoPlayInterval = use(obj.imageAutoPlayInterval, 2000);
        this.imageAutoPlayPauseOnHover = use(obj.imageAutoPlayPauseOnHover, false);
        this.imageInfinityMove = use(obj.imageInfinityMove, false);
        if (obj && obj.imageActions && obj.imageActions.length) {
            obj.imageActions = obj.imageActions.map((/**
             * @param {?} action
             * @return {?}
             */
            action => new NgxGalleryAction(action)));
        }
        this.imageActions = use(obj.imageActions, []);
        this.imageDescription = use(obj.imageDescription, false);
        this.imageBullets = use(obj.imageBullets, false);
        this.thumbnails = use(obj.thumbnails, true);
        this.thumbnailsColumns = use(obj.thumbnailsColumns, 4);
        this.thumbnailsRows = use(obj.thumbnailsRows, 1);
        this.thumbnailsPercent = use(obj.thumbnailsPercent, 25);
        this.thumbnailsMargin = use(obj.thumbnailsMargin, 10);
        this.thumbnailsArrows = use(obj.thumbnailsArrows, true);
        this.thumbnailsArrowsAutoHide = use(obj.thumbnailsArrowsAutoHide, false);
        this.thumbnailsSwipe = use(obj.thumbnailsSwipe, false);
        this.thumbnailsMoveSize = use(obj.thumbnailsMoveSize, 1);
        this.thumbnailsOrder = use(obj.thumbnailsOrder, NgxGalleryOrder.Column);
        this.thumbnailsRemainingCount = use(obj.thumbnailsRemainingCount, false);
        this.thumbnailsAsLinks = use(obj.thumbnailsAsLinks, false);
        this.thumbnailsAutoHide = use(obj.thumbnailsAutoHide, false);
        this.thumbnailMargin = use(obj.thumbnailMargin, 10);
        this.thumbnailSize = use(obj.thumbnailSize, NgxGalleryImageSize.Cover);
        if (obj && obj.thumbnailActions && obj.thumbnailActions.length) {
            obj.thumbnailActions = obj.thumbnailActions.map((/**
             * @param {?} action
             * @return {?}
             */
            action => new NgxGalleryAction(action)));
        }
        this.thumbnailActions = use(obj.thumbnailActions, []);
        this.preview = use(obj.preview, true);
        this.previewDescription = use(obj.previewDescription, true);
        this.previewArrows = use(obj.previewArrows, true);
        this.previewArrowsAutoHide = use(obj.previewArrowsAutoHide, false);
        this.previewSwipe = use(obj.previewSwipe, false);
        this.previewFullscreen = use(obj.previewFullscreen, false);
        this.previewForceFullscreen = use(obj.previewForceFullscreen, false);
        this.previewCloseOnClick = use(obj.previewCloseOnClick, false);
        this.previewCloseOnEsc = use(obj.previewCloseOnEsc, false);
        this.previewKeyboardNavigation = use(obj.previewKeyboardNavigation, false);
        this.previewAnimation = use(obj.previewAnimation, true);
        this.previewAutoPlay = use(obj.previewAutoPlay, false);
        this.previewAutoPlayInterval = use(obj.previewAutoPlayInterval, 2000);
        this.previewAutoPlayPauseOnHover = use(obj.previewAutoPlayPauseOnHover, false);
        this.previewInfinityMove = use(obj.previewInfinityMove, false);
        this.previewZoom = use(obj.previewZoom, false);
        this.previewZoomStep = use(obj.previewZoomStep, 0.1);
        this.previewZoomMax = use(obj.previewZoomMax, 2);
        this.previewZoomMin = use(obj.previewZoomMin, 0.5);
        this.previewRotate = use(obj.previewRotate, false);
        this.previewDownload = use(obj.previewDownload, false);
        this.previewCustom = use(obj.previewCustom, undefined);
        this.previewBullets = use(obj.previewBullets, false);
        this.arrowPrevIcon = use(obj.arrowPrevIcon, 'fa fa-arrow-circle-left');
        this.arrowNextIcon = use(obj.arrowNextIcon, 'fa fa-arrow-circle-right');
        this.closeIcon = use(obj.closeIcon, 'fa fa-times-circle');
        this.fullscreenIcon = use(obj.fullscreenIcon, 'fa fa-arrows-alt');
        this.spinnerIcon = use(obj.spinnerIcon, 'fa fa-spinner fa-pulse fa-3x fa-fw');
        this.zoomInIcon = use(obj.zoomInIcon, 'fa fa-search-plus');
        this.zoomOutIcon = use(obj.zoomOutIcon, 'fa fa-search-minus');
        this.rotateLeftIcon = use(obj.rotateLeftIcon, 'fa fa-undo');
        this.rotateRightIcon = use(obj.rotateRightIcon, 'fa fa-repeat');
        this.downloadIcon = use(obj.downloadIcon, 'fa fa-arrow-circle-down');
        if (obj && obj.actions && obj.actions.length) {
            obj.actions = obj.actions.map((/**
             * @param {?} action
             * @return {?}
             */
            action => new NgxGalleryAction(action)));
        }
        this.actions = use(obj.actions, []);
    }
}
if (false) {
    /** @type {?} */
    NgxGalleryOptions.prototype.width;
    /** @type {?} */
    NgxGalleryOptions.prototype.height;
    /** @type {?} */
    NgxGalleryOptions.prototype.breakpoint;
    /** @type {?} */
    NgxGalleryOptions.prototype.fullWidth;
    /** @type {?} */
    NgxGalleryOptions.prototype.layout;
    /** @type {?} */
    NgxGalleryOptions.prototype.startIndex;
    /** @type {?} */
    NgxGalleryOptions.prototype.linkTarget;
    /** @type {?} */
    NgxGalleryOptions.prototype.lazyLoading;
    /** @type {?} */
    NgxGalleryOptions.prototype.image;
    /** @type {?} */
    NgxGalleryOptions.prototype.imagePercent;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageArrows;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageArrowsAutoHide;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageSwipe;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageAnimation;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageSize;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageAutoPlay;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageAutoPlayInterval;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageAutoPlayPauseOnHover;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageInfinityMove;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageActions;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageDescription;
    /** @type {?} */
    NgxGalleryOptions.prototype.imageBullets;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnails;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsColumns;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsRows;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsPercent;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsMargin;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsArrows;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsArrowsAutoHide;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsSwipe;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsMoveSize;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsOrder;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsRemainingCount;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsAsLinks;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailsAutoHide;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailMargin;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailSize;
    /** @type {?} */
    NgxGalleryOptions.prototype.thumbnailActions;
    /** @type {?} */
    NgxGalleryOptions.prototype.preview;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewDescription;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewArrows;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewArrowsAutoHide;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewSwipe;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewFullscreen;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewForceFullscreen;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewCloseOnClick;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewCloseOnEsc;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewKeyboardNavigation;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewAnimation;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewAutoPlay;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewAutoPlayInterval;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewAutoPlayPauseOnHover;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewInfinityMove;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewZoom;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewZoomStep;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewZoomMax;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewZoomMin;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewRotate;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewDownload;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewCustom;
    /** @type {?} */
    NgxGalleryOptions.prototype.previewBullets;
    /** @type {?} */
    NgxGalleryOptions.prototype.arrowPrevIcon;
    /** @type {?} */
    NgxGalleryOptions.prototype.arrowNextIcon;
    /** @type {?} */
    NgxGalleryOptions.prototype.closeIcon;
    /** @type {?} */
    NgxGalleryOptions.prototype.fullscreenIcon;
    /** @type {?} */
    NgxGalleryOptions.prototype.spinnerIcon;
    /** @type {?} */
    NgxGalleryOptions.prototype.zoomInIcon;
    /** @type {?} */
    NgxGalleryOptions.prototype.zoomOutIcon;
    /** @type {?} */
    NgxGalleryOptions.prototype.rotateLeftIcon;
    /** @type {?} */
    NgxGalleryOptions.prototype.rotateRightIcon;
    /** @type {?} */
    NgxGalleryOptions.prototype.downloadIcon;
    /** @type {?} */
    NgxGalleryOptions.prototype.actions;
}
