// @nomadreservations/ngx-gallery/esm5/lib/ngx-gallery-bullets.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
var NgxGalleryBulletsComponent = /** @class */ (function () {
    function NgxGalleryBulletsComponent() {
        this.active = 0;
        this.changed = new EventEmitter();
    }
    /**
     * @return {?}
     */
    NgxGalleryBulletsComponent.prototype.getBullets = /**
     * @return {?}
     */
    function () {
        return Array(this.count);
    };
    /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    NgxGalleryBulletsComponent.prototype.handleChange = /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    function (event, index) {
        this.changed.emit(index);
    };
    NgxGalleryBulletsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-gallery-bullets',
                    template: "\n    <div\n      class=\"ngx-gallery-bullet\"\n      *ngFor=\"let bullet of getBullets(); let i = index\"\n      (click)=\"handleChange($event, i)\"\n      [ngClass]=\"{ 'ngx-gallery-active': i == active }\"\n    ></div>\n  ",
                    styles: [":host{position:absolute;z-index:2000;display:inline-flex;left:50%;transform:translateX(-50%);bottom:0;padding:10px}.ngx-gallery-bullet{width:10px;height:10px;border-radius:50%;cursor:pointer;background:#fff}.ngx-gallery-bullet:not(:first-child){margin-left:5px}.ngx-gallery-bullet.ngx-gallery-active,.ngx-gallery-bullet:hover{background:#000}"]
                }] }
    ];
    NgxGalleryBulletsComponent.propDecorators = {
        count: [{ type: Input }],
        active: [{ type: Input }],
        changed: [{ type: Output }]
    };
    return NgxGalleryBulletsComponent;
}());
export { NgxGalleryBulletsComponent };
if (false) {
    /** @type {?} */
    NgxGalleryBulletsComponent.prototype.count;
    /** @type {?} */
    NgxGalleryBulletsComponent.prototype.active;
    /** @type {?} */
    NgxGalleryBulletsComponent.prototype.changed;
}
