// @nomadreservations/ngx-gallery/esm5/lib/ngx-gallery-arrows.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
var NgxGalleryArrowsComponent = /** @class */ (function () {
    function NgxGalleryArrowsComponent() {
        this.prevClick = new EventEmitter();
        this.nextClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    NgxGalleryArrowsComponent.prototype.handlePrevClick = /**
     * @return {?}
     */
    function () {
        this.prevClick.emit();
    };
    /**
     * @return {?}
     */
    NgxGalleryArrowsComponent.prototype.handleNextClick = /**
     * @return {?}
     */
    function () {
        this.nextClick.emit();
    };
    NgxGalleryArrowsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-gallery-arrows',
                    template: "\n    <div class=\"ngx-gallery-arrow-wrapper ngx-gallery-arrow-left\">\n      <div\n        class=\"ngx-gallery-icon ngx-gallery-arrow\"\n        aria-hidden=\"true\"\n        (click)=\"handlePrevClick()\"\n        [class.ngx-gallery-disabled]=\"prevDisabled\"\n      >\n        <i class=\"ngx-gallery-icon-content {{ arrowPrevIcon }}\"></i>\n      </div>\n    </div>\n    <div class=\"ngx-gallery-arrow-wrapper ngx-gallery-arrow-right\">\n      <div\n        class=\"ngx-gallery-icon ngx-gallery-arrow\"\n        aria-hidden=\"true\"\n        (click)=\"handleNextClick()\"\n        [class.ngx-gallery-disabled]=\"nextDisabled\"\n      >\n        <i class=\"ngx-gallery-icon-content {{ arrowNextIcon }}\"></i>\n      </div>\n    </div>\n  ",
                    styles: [".ngx-gallery-arrow-wrapper{position:absolute;height:100%;width:1px;display:table;z-index:2000;table-layout:fixed}.ngx-gallery-arrow-left{left:0}.ngx-gallery-arrow-right{right:0}.ngx-gallery-arrow{top:50%;transform:translateY(-50%);cursor:pointer}.ngx-gallery-arrow.ngx-gallery-disabled{opacity:.6;cursor:default}.ngx-gallery-arrow-left .ngx-gallery-arrow{left:10px}.ngx-gallery-arrow-right .ngx-gallery-arrow{right:10px}"]
                }] }
    ];
    NgxGalleryArrowsComponent.propDecorators = {
        prevDisabled: [{ type: Input }],
        nextDisabled: [{ type: Input }],
        arrowPrevIcon: [{ type: Input }],
        arrowNextIcon: [{ type: Input }],
        prevClick: [{ type: Output }],
        nextClick: [{ type: Output }]
    };
    return NgxGalleryArrowsComponent;
}());
export { NgxGalleryArrowsComponent };
if (false) {
    /** @type {?} */
    NgxGalleryArrowsComponent.prototype.prevDisabled;
    /** @type {?} */
    NgxGalleryArrowsComponent.prototype.nextDisabled;
    /** @type {?} */
    NgxGalleryArrowsComponent.prototype.arrowPrevIcon;
    /** @type {?} */
    NgxGalleryArrowsComponent.prototype.arrowNextIcon;
    /** @type {?} */
    NgxGalleryArrowsComponent.prototype.prevClick;
    /** @type {?} */
    NgxGalleryArrowsComponent.prototype.nextClick;
}
