// @nomadreservations/ngx-gallery/esm5/lib/ngx-gallery.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { HammerGestureConfig, HAMMER_GESTURE_CONFIG } from '@angular/platform-browser';
import { NgxGalleryActionComponent } from './ngx-gallery-action.component';
import { NgxGalleryArrowsComponent } from './ngx-gallery-arrows.component';
import { NgxGalleryBulletsComponent } from './ngx-gallery-bullets.component';
import { NgxGalleryImageComponent } from './ngx-gallery-image.component';
import { NgxGalleryPreviewComponent } from './ngx-gallery-preview.component';
import { NgxGalleryThumbnailsComponent } from './ngx-gallery-thumbnails.component';
import { NgxGalleryComponent } from './ngx-gallery.component';
export { NgxGalleryActionComponent } from './ngx-gallery-action.component';
export { NgxGalleryAction } from './ngx-gallery-action.model';
export { NgxGalleryAnimation } from './ngx-gallery-animation.model';
export { NgxGalleryArrowsComponent } from './ngx-gallery-arrows.component';
export { NgxGalleryBulletsComponent } from './ngx-gallery-bullets.component';
export { NgxGalleryHelperService } from './ngx-gallery-helper.service';
export { NgxGalleryImageSize } from './ngx-gallery-image-size.model';
export { NgxGalleryImageComponent } from './ngx-gallery-image.component';
export { NgxGalleryImage } from './ngx-gallery-image.model';
export { NgxGalleryLayout } from './ngx-gallery-layout.model';
export { NgxGalleryOptions } from './ngx-gallery-options.model';
export { NgxGalleryOrder } from './ngx-gallery-order.model';
export { NgxGalleryOrderedImage } from './ngx-gallery-ordered-image.model';
export { NgxGalleryPreviewComponent } from './ngx-gallery-preview.component';
export { NgxGalleryThumbnailsComponent } from './ngx-gallery-thumbnails.component';
export { NgxGalleryComponent } from './ngx-gallery.component';
var CustomHammerConfig = /** @class */ (function (_super) {
    tslib_1.__extends(CustomHammerConfig, _super);
    function CustomHammerConfig() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.overrides = (/** @type {?} */ ({
            pinch: { enable: false },
            rotate: { enable: false }
        }));
        return _this;
    }
    return CustomHammerConfig;
}(HammerGestureConfig));
export { CustomHammerConfig };
if (false) {
    /** @type {?} */
    CustomHammerConfig.prototype.overrides;
}
var NgxGalleryModule = /** @class */ (function () {
    function NgxGalleryModule() {
    }
    NgxGalleryModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [
                        NgxGalleryActionComponent,
                        NgxGalleryArrowsComponent,
                        NgxGalleryBulletsComponent,
                        NgxGalleryImageComponent,
                        NgxGalleryThumbnailsComponent,
                        NgxGalleryPreviewComponent,
                        NgxGalleryComponent
                    ],
                    exports: [NgxGalleryComponent],
                    providers: [{ provide: HAMMER_GESTURE_CONFIG, useClass: CustomHammerConfig }]
                },] }
    ];
    return NgxGalleryModule;
}());
export { NgxGalleryModule };
